cli/src/complete.rs

```diff
@@ -354,6 +354,244 @@ pub fn leaf_config_keys() -> Vec<CompletionCandidate> {
+        for case in bad_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(parse_revision_impl(args), None, "case: {case:?}");
+        }
+    }
+
+    pub fn revision() -> Option<String> {
+        parse_revision_impl(std::env::args())
+    }
+
+    pub fn revision_or_wc() -> String {
+        revision().unwrap_or_else(|| "@".into())
+    }
+
+    fn parse_range_impl<T>(args: impl Fn() -> T) -> Option<(String, String)>
+    where
+        T: Iterator<Item = String>,
+    {
+        let from = parse_flag(&["-f", "--from"], args())?;
+        let to = parse_flag(&["-t", "--to"], args()).unwrap_or_else(|| "@".into());
+
+        Some((from, to))
+    }
+
+    #[test]
+    fn test_parse_range_impl() {
+        let wc_cases: &[&[&str]] = &[
+            &["-f", "foo"],
+            &["--from", "foo"],
+            &["-f=foo"],
+            &["preceding_arg", "-f", "foo"],
+            &["-f", "foo", "following_arg"],
+        ];
+        for case in wc_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(
+                parse_range_impl(|| args.clone()),
+                Some(("foo".into(), "@".into())),
+                "case: {case:?}",
+            );
+        }
+        let to_cases: &[&[&str]] = &[
+            &["-f", "foo", "-t", "bar"],
+            &["-f", "foo", "--to", "bar"],
+            &["-f=foo", "-t=bar"],
+            &["-t=bar", "-f=foo"],
+        ];
+        for case in to_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(
+                parse_range_impl(|| args.clone()),
+                Some(("foo".into(), "bar".into())),
+                "case: {case:?}",
+            );
+        }
+        let bad_cases: &[&[&str]] = &[&[], &["-f"], &["foo"], &["-R", "foo"], &["-R=foo"]];
+        for case in bad_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(parse_range_impl(|| args.clone()), None, "case: {case:?}");
+        }
+    }
+
+    pub fn range() -> Option<(String, String)> {
+        parse_range_impl(std::env::args)
+    }
+
+    // Special parse function only for `jj squash`. While squash has --from and
+    // --to arguments, only files within --from should be completed, because
+    // the files changed only in some other revision in the range between
+    // --from and --to cannot be squashed into --to like that.
+    pub fn squash_revision() -> Option<String> {
+        if let Some(rev) = parse_flag(&["-r", "--revision"], std::env::args()) {
+            return Some(rev);
+        }
+        parse_flag(&["-f", "--from"], std::env::args())
+    }
+}
+
+fn all_files_from_rev(rev: String) -> Vec<CompletionCandidate> {
+    with_jj(|mut jj, _| {
+        let output = jj
+            .arg("file")
+            .arg("list")
+            .arg("--revision")
+            .arg(rev)
+            .output()
+            .map_err(user_error)?;
```

Comment:
I think the most elegant solution is the one you suggested [here](https://github.com/martinvonz/jj/pull/4895#discussion_r1845728977):

> We also need to give `jj file list` (or `jj debug tree`?) an option to also list directory entries so we don't have to list all files in subdirectories.

If that was implemented, we could even generate completions even if the user hasn't typed anything yet, right? Assuming that listing all _directories_ at google is not also extremely slow.

I don't mind updating this PR, I'm just not sure yet what's the best change to put where / in which PR.

Reply:
Btw clap doesn't handle it, I tried to complete the file list of `all()` revisions of the jj repo while working on this and it took 9 seconds, so clap doesn't seem to limit the time a completion is allowed to take.

Reply:
> If that was implemented, we could even generate completions even if the user hasn't typed anything yet, right? 

Yes.

> Assuming that listing all directories at google is not also extremely slow.

It won't be extremely slow, but we do have directories with way more matches than useful. But that's a simpler problem to solve (just truncate it or something).

Reply:
> I don't mind updating this PR, I'm just not sure yet what's the best change to put where / in which PR.

I'm not sure either. I think it would be helpful at Google (and for Linux and nixpkgs devs, probably) if simply typing `jj diff <TAB>` doesn't hang. It might be that shell completion will still be too slow in other cases that we can't enable it anyway, but that seems like a good start at least.

Reply:
For now, I made it so completions are skipped completely if the user hasn't started typing anything yet and the output is capped to 1k lines.

(I tested that the output is limited and returns quickly by replacing the `jj` command with `yes`, which worked as expected.)